source/opt/passes.h

```diff
@@ -74,6 +80,142 @@ class EliminateDeadConstantPass : public Pass {
   bool Process(ir::Module*) override;
 };
 
+// The optimization pass for folding SpecConstantOp and SpecConstantComposite
+// instructions whose elements are all known normal constants.
+class FoldSpecConstantOpAndCompositePass : public Pass {
+ public:
+  FoldSpecConstantOpAndCompositePass()
+      : max_id_(0),
+        def_use_mgr_(nullptr),
+        type_mgr_(nullptr),
+        id_to_const_val_() {}
+  const char* name() const override { return "fold-spec-const-op-composite"; }
+  bool Process(ir::Module* module) override {
+    Initialize(module);
+    return ProcessImpl(module);
+  };
+
+ private:
+  // Initializes the type manager, def-use manager and get the maximal id used
+  // in the module.
+  void Initialize(ir::Module* module) {
+    type_mgr_.reset(new analysis::TypeManager(*module));
+    def_use_mgr_.reset(new analysis::DefUseManager(module));
+    for (const auto& id_def : def_use_mgr_->id_to_defs()) {
```

Comment:
Funny.  I have a MR in flight that computes this (without a def use manager).  But I can refactor that later.